NCI accessors: chunk_edge (normal, intersection_factor) and point (transfer_function)

int-ptr-ptr opened the pull request:
## Description

Please describe the changes/features in this pull request.
- updates the point `impl_load` for nonconforming interfaces
- adds new `chunk_edge` containers for `compute_coupling`
  - since different flux schemes need different data, there is no one-size-fits-all `interface_data` container. We can decide how to handle delegation along with the accessor compatibility update we will have to do in the future for coupled interfaces.

## Issue Number

If there is an issue created for these changes, link it here

## Checklist

Please make sure to check developer documentation on specfem docs.

- [ ] I ran the code through pre-commit to check style
- [ ] **THE DOCUMENTATION BUILDS WITHOUT WARNINGS/ERRORS**
- [ ] I have added labels to the PR (see right hand side of the PR page)
- [ ] My code passes all the integration tests
- [ ] I have added sufficient unittests to test my changes
- [ ] I have added/updated documentation for the changes I am proposing
- [ ] I have updated CMakeLists to ensure my code builds
- [ ] My code builds across all platforms

int-ptr-ptr committed:
additional accessors necessary for dg2d kernel

Rohit-Kakodkar commented on core/specfem/point/impl/nonconforming_transfer_function.hpp:
```diff
@@ -0,0 +1,157 @@
+  TransferViewType transfer_function_self;
+  KOKKOS_INLINE_FUNCTION const TransferViewType &get_transfer_function() const {
+    return transfer_function_self;
+  }
```
We dont need the get function since `transfer_function_self` is itself public.

int-ptr-ptr replied:
`get_transfer_function()` is used to abstract the difference between `transfer_function_self` and `transfer_function_coupled` so that the same code can be used.

As an example, [here's](https://github.com/int-ptr-ptr/SPECFEMPP/blob/dg2d-kernelimpl/include/algorithms/transfer.hpp#L88) the transfer algorithm using it, where `interface_data` has already been checked to inherit only one of `transfer_function<is_self=true>` and `transfer_function<is_self=false>`. `get_transfer_function()` is used instead of something like
```
(stores_self_transfer_function<decltype(interface_data)>::type)?
    interface_data.transfer_function_self : interface_data.transfer_function_coupled
```

Would storing a `const TransferViewType&` field be better than the function?

Rohit-Kakodkar replied:
I am not getting this. What cant you directly reference it in the line you linked `const auto& transfer_functions = interface_data.transfer_function`

int-ptr-ptr replied:
> I am not getting this. What cant you directly reference it in the line you linked `const auto& transfer_functions = interface_data.transfer_function`

`transfer_function_self` is different from `transfer_function_coupled`. You would need a compile-time ternary, which is what the `get_transfer_function()` method is.

int-ptr-ptr replied:
https://github.com/PrincetonUniversity/SPECFEMPP/blob/68e7089ae1a38aaedaac62e606c124d2a3db0679/core/specfem/point/impl/nonconforming_transfer_function.hpp#L57-L62:

https://github.com/PrincetonUniversity/SPECFEMPP/blob/68e7089ae1a38aaedaac62e606c124d2a3db0679/include/algorithms/transfer.hpp#L108-L111

Is this what you had in mind?

Rohit-Kakodkar commented on core/specfem/point/impl/nonconforming_transfer_function.hpp:
```diff
@@ -0,0 +1,157 @@
+   *
+   * @return int Amount of shared memory in bytes
+   */
+  constexpr static int shmem_size() { return TransferViewType::shmem_size(); }
```
Point views would not use a team allocator. Point views are instantiated in register space and not scrach space. For example, `shmem_size()` does not exist for `datatype::VectorPointViewType`

Rohit-Kakodkar commented on core/specfem/point/impl/nonconforming_transfer_function.hpp:
```diff
@@ -0,0 +1,157 @@
+   * enabled.
+   */
+  TransferViewType transfer_function_coupled;
+  KOKKOS_INLINE_FUNCTION const TransferViewType &get_transfer_function() const {
```
See comment above.

Rohit-Kakodkar commented on core/specfem/point/impl/nonconforming_transfer_function.hpp:
```diff
@@ -0,0 +1,157 @@
+   *
+   * @return int Amount of shared memory in bytes
+   */
+  constexpr static int shmem_size() { return TransferViewType::shmem_size(); }
```
See comment above.

Rohit-Kakodkar commented on core/specfem/assembly/nonconforming_interfaces/dim2/impl/nonconforming_interface.hpp:
```diff
@@ -121,20 +121,46 @@ struct interface_container<specfem::dimension::type::dim2, InterfaceTag,
                 specfem::data_access::is_point<PointType>::value, int> = 0>
   KOKKOS_FORCEINLINE_FUNCTION void impl_load(const IndexType &index,
                                              PointType &point) const {
-    if constexpr (on_device) {
```
Was this wrong before?

int-ptr-ptr replied:
It was outdated and never had to be compiled, since no `point<connection_tag=nonconforming>` accessor existed.

Rohit-Kakodkar replied:
Can you merge branch `issue-1350` into this. That branch should provide a fix for this.

Rohit-Kakodkar reviewed (changes requested):
Small changes.

lsawade reviewed (approved):
LGTM

int-ptr-ptr committed:
removed `shmem_size` from point accessors

int-ptr-ptr committed:
replaced get_transfer_function method with reference

int-ptr-ptr committed:
Merge branch 'devel' into nci-accessor

int-ptr-ptr committed:
gpu not compiling again -- points to ncmarked_conforming_grid_flat

int-ptr-ptr committed:
Merge branch 'devel' into nci-accessor

int-ptr-ptr committed:
modified cmakelists

Rohit-Kakodkar commented on core/specfem/chunk_edge/impl/nonconforming_transfer_function.hpp:
```diff
@@ -64,9 +64,7 @@ struct nonconforming_transfer_function<
-  KOKKOS_INLINE_FUNCTION const TransferViewType &get_transfer_function() const {
-    return transfer_function_self;
-  }
+  TransferViewType &transfer_function = transfer_function_self;
```
I still dont understand why you need a reference type? Can you not name the above variable just `transfer_function`?

int-ptr-ptr replied:
There will be a flux scheme where `compute_coupling` needs both `transfer_function_self` and `transfer_function_coupled`, and it makes more sense to do it this way than to do something like
`intersection_data.specfem::chunk_edge::nonconforming_transfer_function<true,...>::transfer_function`.

Right now, the compatibility checker looks for the field names, so if this is something to change, I would rather wait until we rework that.

Rohit-Kakodkar replied:
Would updating this to `const auto &transfer_function = interface_container.transfer_function_self` in the transfer kernel do the trick?

Rohit-Kakodkar commented on core/specfem/point/impl/nonconforming_transfer_function.hpp:
```diff
@@ -0,0 +1,140 @@
+   */
+  template <typename MemberType>
+  KOKKOS_INLINE_FUNCTION nonconforming_transfer_function(const MemberType &team)
+      : transfer_function_self(team.team_scratch(0)) {}
```
Point views should not have team operator.

Rohit-Kakodkar commented on core/specfem/point/impl/nonconforming_transfer_function.hpp:
```diff
@@ -0,0 +1,140 @@
+   */
+  template <typename MemberType>
+  KOKKOS_INLINE_FUNCTION nonconforming_transfer_function(const MemberType &team)
+      : transfer_function_coupled(team.team_scratch(0)) {}
```
Same as above

Rohit-Kakodkar reviewed (commented):
See comments.

Rohit-Kakodkar commented on core/specfem/point/impl/nonconforming_transfer_function.hpp:
```diff
@@ -0,0 +1,140 @@
+          specfem::dimension::type DimensionTag, bool UseSIMD>
+struct nonconforming_transfer_function;
+
+template <int NQuadIntersection, bool UseSIMD>
```
I dont think we'd be able to SIMD vectorize compute coupling yet just because of how data is layed out.

Rohit-Kakodkar committed:
Fix bug in coupled interfaces

- [x] Update index function in edge/intersection iterator
- [x] Update `for_all` execution function

int-ptr-ptr committed:
removed scratch constructor from point

int-ptr-ptr committed:
removed simd

int-ptr-ptr committed:
removed simd from test -- compiles now

Rohit-Kakodkar committed:
compiler fix

int-ptr-ptr committed:
polarized transfer algorithm, removed ambiguated transfer_function ref

Rohit-Kakodkar committed:
Merge branch 'devel' into issue-1350

int-ptr-ptr committed:
Merge remote-tracking branch 'public_repo/issue-1350' into nci-accessor

int-ptr-ptr committed:
ncmarked_conforming_grid_curved: rel src

revert  ncmarked_conforming_grid_curved CMakeLists.txt sources to relative (to match ncmarked_conforming_grid_flat CmakeLists)
